internal/stacks/stackManager.go

```diff
@@ -0,0 +1,597 @@
+	}
+	return s.ensureDirectories()
+}
+
+func (s *StackManager) RemoveStack(verbose bool) error {
+	if err := s.ResetStack(verbose); err != nil {
+		return err
+	}
+	return os.RemoveAll(filepath.Join(constants.StacksDir, s.Stack.Name))
+}
+
+func (s *StackManager) checkPortsAvailable() error {
+	ports := make([]int, 1)
+	ports[0] = s.Stack.ExposedBlockchainPort
+	for _, member := range s.Stack.Members {
+		ports = append(ports, member.ExposedDataexchangePort)
+		ports = append(ports, member.ExposedEthconnectPort)
+		if !member.External {
+			ports = append(ports, member.ExposedFireflyAdminPort)
+			ports = append(ports, member.ExposedFireflyPort)
+		}
+		ports = append(ports, member.ExposedIPFSApiPort)
+		ports = append(ports, member.ExposedIPFSGWPort)
+		ports = append(ports, member.ExposedPostgresPort)
+		ports = append(ports, member.ExposedUIPort)
+		ports = append(ports, member.ExposedTokensPort)
+	}
+	for _, port := range ports {
+		available, err := checkPortAvailable(port)
+		if err != nil {
+			return err
+		}
+		if !available {
+			return fmt.Errorf("port %d is unavailable. please check to see if another process is listening on that port", port)
+		}
+	}
+	return nil
+}
+
+func checkPortAvailable(port int) (bool, error) {
+	timeout := time.Millisecond * 500
+	conn, err := net.DialTimeout("tcp", net.JoinHostPort("127.0.0.1", fmt.Sprint(port)), timeout)
+
+	if netError, ok := err.(net.Error); ok && netError.Timeout() {
+		return true, nil
+	}
+
+	switch t := err.(type) {
+
+	case *net.OpError:
+		switch t := t.Unwrap().(type) {
+		case *os.SyscallError:
+			if t.Syscall == "connect" {
+				return true, nil
+			}
+		}
+		if t.Op == "dial" {
+			return false, err
+		} else if t.Op == "read" {
+			return true, nil
+		}
+
+	case syscall.Errno:
+		if t == syscall.ECONNREFUSED {
+			return true, nil
+		}
+	}
+
+	if conn != nil {
+		defer conn.Close()
+		return false, nil
+	}
+	return true, nil
+}
+
+func (s *StackManager) runFirstTimeSetup(verbose bool, options *StartOptions) error {
+	workingDir := filepath.Join(constants.StacksDir, s.Stack.Name)
+
+	s.Log.Info("initializing blockchain node")
+	if err := s.blockchainProvider.RunFirstTimeSetup(); err != nil {
+		return err
+	}
+
+	s.Log.Info("writing data exchange certs")
+	if err := s.writeDataExchangeCerts(verbose); err != nil {
+		return err
+	}
+
+	// write firefly configs to volumes
+	for _, member := range s.Stack.Members {
+		if !member.External {
+			s.Log.Info(fmt.Sprintf("copying firefly.core to firefly_core_%s", member.ID))
+			volumeName := fmt.Sprintf("%s_firefly_core_%s", s.Stack.Name, member.ID)
+			if err := docker.CopyFileToVolume(volumeName, path.Join(workingDir, "configs", fmt.Sprintf("firefly_core_%s.yml", member.ID)), "/firefly.core", verbose); err != nil {
+				return err
+			}
+		}
+	}
+
+	if !options.NoPull {
+		s.Log.Info("pulling latest versions")
+		if err := docker.RunDockerComposeCommand(workingDir, verbose, verbose, "pull"); err != nil {
+			return err
+		}
+	}
+
```

Comment:
Is it intentional to run `PostStart` but not `PreStart` in this path? Should some of the common "start" logic for first-time and not-first-time be factored out? There's code duplication between here and line ~318.

Reply:
Let me double check this... the way it is intended to work is that `PreStart` and `PostStart` run every time. `RunFirstTimeSetup` and `DeploySmartContracts` only run the first time.

Reply:
K, seems like maybe `PreStart` + `start dockers` + `PostStart` + `ensure up` could be factored out to a common utility method, since those 4 things happen in both paths.

Reply:
Good catch on the code duplication. I just pushed another commit that addresses that.